### ISSUE TITLE
Why hidapi and not libhid?

### ISSUE THREAD

**Issue by opoplawski** [edorfaus/TEMPered#2]
hidapi is not really packaged very well for a library.  Why use it over say libhid?  Cross-platform support?

**Comment by edorfaus** [comment]
I don't quite remember why, but I think it was simply the best-looking of the HID implementations I found when I went looking for one when I started this project.

I'm not sure what you mean by well packaged here, but if you don't think it is packaged well enough, why not help them get it packaged the way you think it should be? They do accept patches (I sent one that was accepted)...

Now that I think about it, I could probably make this a bit easier for my users by including hidapi as a submodule in my git repo... It's linked statically anyway. I'm pretty new to git though, so haven't actually done that before, and wasn't really even aware of it back when I started... What do you think? Should I do that?

As for libhid, which one do you mean? A quick googling found me at least two different projects using that name, where the first one (on alioth) was last updated in 2008, uses libusb0.1, and is using the GPLv2; and the other one is written in C++ (aaand... doesn't appear to support Linux now I that look for it)...

**Comment by opoplawski** [comment]
Well, hidapi seems to just produce a single .o.  It doesn't produce a library and header for install so it can't really be packaged in this state.  I'll contact them though.

It won't help me in this case to have hidapi as a submodule as that's not allowed in Fedora packaging, which is what I'm looking at doing.

http://libhid.alioth.debian.org/ is what is in Fedora.  Have not looked at it other than to see that it is already packaged.

**Comment by edorfaus** [comment]
Ah, I see. I think hidapi is mainly intended for static linking, for which you only need the .o and .h while building the program - but I can see how that can be a bit of a problem for distro packagers.
(Actually - looking at the readme, it seems they've intended for developers to copy the .c into their own project... Can't say I really like that approach, simpler though it may be, as it makes it harder to grab new versions of the library.)

It does have a header file though - it's not generated, as it's the same for all variants, but it's there: hidapi/hidapi.h

For the .a (or maybe you want a .so?) there's another minor snag for Linux, in that there are two implementations to choose from, one based on libusb1, and one on the kernel's hidraw interface. While the libusb1 implementation is default, I tend to use the hidraw one myself... Not sure how to best handle that, easiest is probably to just pick an implementation, not giving any choice to those who just install it. (I guess one way would be to provide two conflicting packages that both provide libhidapi-dev (or whatever you'll call the package), perhaps? Or maybe I'm just too used to how Debian sometimes does things...)

Generating the .a file from the compiled .o shouldn't be hard though, at least it wasn't for me when I added that to my project - just a couple of standard commands in the makefile target. I haven't really looked into building a .so yet though.

**Comment by edorfaus** [comment]
Hm, I have a question. Well, several actually. (Though mostly for curiosity.)

If I had done like hidapi's readme suggests, and copied the hid.c (and hidapi.h) file into my own project, and built it as part of the normal make process - thus linking it into my own library - would that have been allowed in Fedora packaging? Or would you have had to split it out and build hidapi separately anyway?

If that is allowed, would it really be any different if I integrated it as a submodule, made my makefile build it automatically, and included the resulting object in my .a and .so files?

I'm not so sure I really want to do either of those things, but I might be persuaded to do the latter if it would make things easier (e.g. if getting upstream to listen is difficult)... Of course, if it _wouldn't_ make anything easier, I probably won't.

By the way, I added Makefile targets to build a shared library and programs that use it - and it seems to work for me, but I don't really know what to check for. Do you see any problems with how I did this?

Oh, and, just FYI, I believe I have limited fairly well which parts of the code are directly affected by hidapi, so I think there should only be one or two files that would need changing to use a different HID library. In fact, I think it should be feasible to support multiple alternative implementations, kinda like hidapi does... I don't think I'm personally interested in doing the work needed to use another HID library (I have enough to do already), but if someone else does it, I might include the result.

**Comment by opoplawski** [comment]
Might not be detected, but wouldn't be allowed if it was.

There was a submission for shared lib support for hidapi about 3 months ago.  I commented on that.

I may take a look at porting to libhid then.

**Comment by edorfaus** [comment]
Ah, OK. I understand.

Yeah, I saw that comment... I've been considering whether I should chime in on it too, or just submit my own (pretending I didn't see the existing one or something), in an attempt to get signal11's attention on the issue. If I do the latter, I could include the fixed soname and stuff... (Assuming I've understood it right. :P )

As for the porting, I'm currently doing some relatively major reworking that touches that part of the code (to support multiple sensors per device), so you may want to hold off a bit on actually doing that porting, to avoid wasting effort.

It's up to you of course; I'm honestly rather humbled that someone even wants to package my little project for their distro...

Oh, and I just thought of another thing to consider... I'm currently using the BSD 2-clause license, and would prefer to keep doing so. IANAL, so I'm not really sure how that would be affected by including a port to using a GPLv2 library (which libhid is AFAIK) - but as far as I understand, it would make the whole project GPLv2 when linked with the code for that library, wouldn't it?

However, I also think that it could still be used purely under the BSD license, when built without that port, just so long as that build would still be useful - meaning as long as I keep the current implementation (or another BSD one) working.

Which means that I can't really replace the current implementation, just add an alternative to it, if I want to keep the BSD license.

Do you know if this interpretation is correct?

**Comment by opoplawski** [comment]
Okay, I'll wait for your code reorg to take a look at libhid.  Maybe hidapi will have a better build by then too.  Anything to poke them along would be good.

As for licensing - you cannot release your program as it is with the BSD license because you are directly combining it with GPL licensed code.  If hidapi  was a separate library that you linked to, then you would be okay.

**Comment by edorfaus** [comment]
Uhm... First off, I don't really see how linking to a .o file from hidapi is materially different from linking to a .so or .a file from hidapi. I'm even being explicit both in instructions and the Makefile that it has to built separately, so I figured I was basically using it as a separate library that I'm linking to. As far as I understand though, that doesn't really matter for the GPL's purposes, which says you're linking to it whichever of those ways you're using.

Secondly, and more importantly I believe, I'm not using hidapi under the GPL, I'm using it under the BSD-style license (see hidapi's LICENSE.txt and LICENSE-bsd.txt files). At least that is my intent - do you see something stopping me from doing that?

**Comment by opoplawski** [comment]
Ah, didn't realize that you were talking about libhid and GPL.  Yeah, with hidapi being multilicensed that is fine.  But combining source files together (multiple .o s) is different than linking to an external library.

**Comment by edorfaus** [comment]
HIDAPI now builds a shared library (two, actually), and I've tested that they work with this project.

My Makefile needs a patch to actually use it, but I'll take a look at doing that properly soon (meanwhile, it works to set it in HIDAPI_OBJ and appending the -L and -l flags to LIBS in the else, instead of the udev stuff).

By the way, I'm a bit unsure about something related to this. Do you think I should make my .so depend on the hidapi library, so it gets pulled in automatically? If I do, programs won't have to link explicitly with hidapi like they do now, but if I don't, programs can choose which hidapi implementation they want to link with...

**Comment by opoplawski** [comment]
On 03/29/2012 07:27 PM, Frode Austvik wrote:

> HIDAPI now builds a shared library (two, actually), and I've tested that they work with this project.
> 
> My Makefile needs a patch to actually use it, but I'll take a look at doing that properly soon (meanwhile, it works to set it in HIDAPI_OBJ and appending the -L and -l flags to LIBS in the else, instead of the udev stuff).
> 
> By the way, I'm a bit unsure about something related to this. Do you think I should make my .so depend on the hidapi library, so it gets pulled in automatically? If I do, programs won't have to link explicitly with hidapi like they do now, but if I don't, programs can choose which hidapi implementation they want to link with...

Yeah, that's a funny one, but I think not linking directly to one is warranted 
in this case.  I managed to package up the new hidapi so that looks good. 
Thanks for the heads up.

## 

Orion Poplawski
Technical Manager                     303-415-9701 x222
NWRA, Boulder Office                  FAX: 303-415-9702
3380 Mitchell Lane                       orion@nwra.com
Boulder, CO 80301                   http://www.nwra.com

**Comment by edorfaus** [comment]
NP, yw.

Just FYI, I'm also done with the reworking now (it's merged to master). I'm still thinking about how to best enable use of the hidapi shared libs in my Makefile; it's actually starting to get complex enough (esp. considering hidapi might be installed on the system instead of just checked out nearby) that I probably should look into a real build system, with a ./configure (or equivalent) and everything...

**Comment by opoplawski** [comment]
On 04/06/2012 04:47 PM, Frode Austvik wrote:

> NP, yw.
> 
> Just FYI, I'm also done with the reworking now (it's merged to master). I'm still thinking about how to best enable use of the hidapi shared libs in my Makefile; it's actually starting to get complex enough (esp. considering hidapi might be installed on the system instead of just checked out nearby) that I probably should look into a real build system, with a ./configure (or equivalent) and everything...

Yeah, I would recommend that.  Check out cmake if you haven't already.  It's 
really not to bad and is cross platform.

## 

Orion Poplawski
Technical Manager                     303-415-9701 x222
NWRA, Boulder Office                  FAX: 303-415-9702
3380 Mitchell Lane                       orion@nwra.com
Boulder, CO 80301                   http://www.nwra.com

**Comment by opoplawski** [comment]
On 04/06/2012 04:47 PM, Frode Austvik wrote:

> NP, yw.
> 
> Just FYI, I'm also done with the reworking now (it's merged to master). I'm still thinking about how to best enable use of the hidapi shared libs in my Makefile; it's actually starting to get complex enough (esp. considering hidapi might be installed on the system instead of just checked out nearby) that I probably should look into a real build system, with a ./configure (or equivalent) and everything...
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/edorfaus/TEMPered/issues/2#issuecomment-5003887

Did you push the changes?  the github master seems a little old.

## 

Orion Poplawski
Technical Manager                     303-415-9701 x222
NWRA, Boulder Office                  FAX: 303-415-9702
3380 Mitchell Lane                       orion@nwra.com
Boulder, CO 80301                   http://www.nwra.com

**Comment by edorfaus** [comment]
Yeah, I did - it's just that the merge and push happened a couple days before I mentioned it here (so about 5 days ago now), I just forgot to mention it here until your response reminded me about it.

As for CMake, yeah, I had started looking into it already, and I like what I've seen so far - thanks for the suggestion though, it helps to know others like it too. In fact, if my like of it continues (and so far, I have no reason to think it won't), I'll probably end up merging a build system based on it fairly soon, which should default to building dynamically linked binaries. :)
